File: 286-Walls and Gates.py
from collections import deque
class Solution(object):
    def wallsAndGates(self, rooms):
        """
        :type rooms: List[List[int]]
        :rtype: None Do not return anything, modify rooms in-place instead.
        """
# time complexity: O(mn)
        if not rooms or not rooms[0]:
            return
        
        row = len(rooms)
        col = len(rooms[0])
        
        queue = deque([])
        
        for r in range(row):
            for c in range(col):
                if rooms[r][c] == 0:
                    queue.append((r,c,0))
                    
        while queue:
            size = len(queue)
            for _ in range(size):
                currR, currC, step = queue.popleft()
                for dr, dc in ((1,0),(0,1),(-1,0),(0,-1)):
                    newR = currR + dr
                    newC = currC + dc
                    if 0<=newR<row and 0<=newC<col and rooms[newR][newC] == 2147483647:
                        rooms[newR][newC] = step+1
                        queue.append((newR, newC, step+1))

File: C++/286-Walls and Gates.cpp
class Solution {
public:
    int d[5] = {1, 0, -1, 0, 1};
    void wallsAndGates(vector<vector<int>>& rooms) {
        int row = rooms.size(), col = rooms[0].size();
        int INF = 2147483647;
        queue<pair<int, int>> q;

        for (int r=0; r<row; r++) {
            for (int c=0; c<col; c++) {
                if (rooms[r][c] == 0) {
                    q.push({r, c});
                }
            }
        }


        int step = 1;
        while (!q.empty()) {
            int sz = q.size();
            while (sz--) {
                auto [currR, currC] = q.front();
                q.pop();
                for (int i=0; i<4; i++) {
                    int nr = currR + d[i], nc = currC + d[i+1];
                    if (nr>=0 && nr<row && nc>=0 && nc<col && rooms[nr][nc] == INF) {
                        rooms[nr][nc] = step;
                        q.push({nr, nc});
                    }
                }
            }
            step++;
        }
    }
};